src/pe_crypto.cpp

```diff
@@ -101,85 +103,99 @@ namespace libtorrent
 		m_xor_mask = h.final();
 	}
 
-	int encryption_handler::encrypt(std::vector<boost::asio::mutable_buffer>& iovec)
+	std::tuple<int, std::vector<boost::asio::const_buffer>>
+	encryption_handler::encrypt(
+		aux::array_view<boost::asio::mutable_buffer> iovec)
 	{
+		using namespace boost::asio;
+
 		TORRENT_ASSERT(!m_send_barriers.empty());
 		TORRENT_ASSERT(m_send_barriers.front().enc_handler);
 
 		int to_process = m_send_barriers.front().next;
 
+		boost::asio::mutable_buffer* bufs;
+		int num_bufs;
+		bool need_destruct = false;
 		if (to_process != INT_MAX)
 		{
-			for (std::vector<boost::asio::mutable_buffer>::iterator i = iovec.begin();
-				to_process >= 0; ++i)
+			bufs = TORRENT_ALLOCA(mutable_buffer, iovec.size());
+			need_destruct = true;
+			num_bufs = 0;
+			for (int i = 0; to_process > 0; ++i)
 			{
-				if (to_process == 0)
+				++num_bufs;
+				int const size = buffer_size(iovec[i]);
+				if (to_process < size)
 				{
-					iovec.erase(i, iovec.end());
-					break;
+					new (&bufs[i]) mutable_buffer(
+						buffer_cast<void*>(iovec[i]), to_process);
+					to_process = 0;
 				}
-				else if (to_process < boost::asio::buffer_size(*i))
+				else
 				{
-					*i = boost::asio::mutable_buffer(boost::asio::buffer_cast<void*>(*i), to_process);
-					iovec.erase(++i, iovec.end());
-					to_process = 0;
-					break;
+					new (&bufs[i]) mutable_buffer(iovec[i]);
+					to_process -= size;
 				}
-				to_process -= int(boost::asio::buffer_size(*i));
 			}
-			TORRENT_ASSERT(to_process == 0);
 		}
-
-#if TORRENT_USE_ASSERTS
-		to_process = 0;
```

Comment:
as far as I can see, the re-use of the "to_process" variable here is only used further down, in the next TORRENT_USE_ASSERTS block. I moved this down there, and also used a new variable, to not accidentally mix them or their scope up in the future